copier/vcs.py

```diff
@@ -108,23 +111,68 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
         return latest_tag
 
 
-def clone(url: str, ref: OptStr = None) -> str:
-    """Clone repo into some temporary destination.
+def _perform_clone(src: str, dst: str, ref: OptStr = None) -> str:
+    """Clone from src into dst.
 
     Args:
-        url:
-            Git-parseable URL of the repo. As returned by
-            [get_repo][copier.vcs.get_repo].
+        src:
+            Git-parseable repo clone source.
+        dst:
+            Git-parseable repo clone destination.
         ref:
             Reference to checkout. For Git repos, defaults to `HEAD`.
     """
-    location = tempfile.mkdtemp(prefix=f"{__name__}.clone.")
-    _clone = git["clone", "--no-checkout", url, location]
+    _clone = git["clone", "--no-checkout", src, dst]
     # Faster clones if possible
     if GIT_VERSION >= Version("2.27"):
         _clone = _clone["--filter=blob:none"]
     _clone()
-    with local.cwd(location):
+
+    with local.cwd(dst):
         git("checkout", ref or "HEAD")
         git("submodule", "update", "--checkout", "--init", "--recursive", "--force")
-    return location
+
+    return dst
+
+
+def clone(url: str, ref: OptStr = None) -> str:
+    """Clone repo into some temporary destination.
+
+    Includes dirty changes for local templates by copying into a temp
+    directory and applying a wip commit there.
+
+    Args:
+        url:
+            Git-parseable URL of the repo. As returned by
+            [get_repo][copier.vcs.get_repo].
+        ref:
+            Reference to checkout. For Git repos, defaults to `HEAD`.
+    """
+
+    location = tempfile.mkdtemp(prefix=f"{__name__}.clone.")
+
+    if os.path.exists(url) and Path(url).is_dir():
+        is_dirty = False
+        with local.cwd(url):
+            is_dirty = bool(git("status", "--porcelain").strip())
+        if is_dirty:
+            url_abspath = Path(url).absolute()
+            with TemporaryDirectory(prefix=f"{__name__}.dirty.") as src:
+                with local.cwd(src):
+                    git("--git-dir=.git", f"--work-tree={url_abspath}", "init")
+                    git("--git-dir=.git", f"--work-tree={url_abspath}", "add", "-A")
+                    git(
```

Comment:
This is a smart solution!

Have you tried this?:

1. Clone.
2. Do this same hack, but on the cloned directory.

Reply:
This is possible, but cloning a dirty git-tracked template seems more error prone to me. For example, it may not have a HEAD. I believe the submodule update would also have to be avoided and then called after the hack.

Reply:
Ok, but in that case I still have some doubts about if this will work fine for updates, because if I'm understanding the code correctly, this is not copying the git history correctly, so it wouldn't work as expected for updates. So please could you update the test? (See below)

Note: git has no problem in cloning empty repos AFAIK.

Reply:
yeah, you're totally right. had to move things around a little bit to account for submodules, but this looks to be working now after cloning first